README.md
# CH1 XHR

[![CircleCI](https://circleci.com/gh/bennett000/ch1-xhr.svg?style=svg)](https://circleci.com/gh/bennett000/ch1-xhr)

_This is not well maintained_

## Installation

`yarn add @ch1/xhr`

## License

[LGPL](./LICENSE 'Lesser GNU Public License')


CHANGELOG.md
# CHANGE LOG

## 0.3.0

- Ships as es6 and cjs modules

## 0.2.0

- Ships as es6 modules


src/xhr.spec.ts
import { noop } from '@ch1/utility';
import { Xhr, parseQuery } from './xhr';

describe('xhr wrapper', () => {
  describe('Xhr constructor function', () => {
    let xhrObj: XMLHttpRequest;

    function FakeHttpRequest() {
      return xhrObj;
    }

    beforeEach(() => {
      xhrObj = ({
        abort: noop,
        open: noop,
        readyState: -1,
        setRequestHeader: noop,
        send: noop,
        status: -1
      } as any) as XMLHttpRequest; // yes, this is fake :)
    });

    it('creates with a custom ', () => {
      const x = Xhr(FakeHttpRequest);

      expect(x).not.toBe(undefined);
    });

    it('times out if nothing happens', () => {
      const x = Xhr(FakeHttpRequest, 1);

      return expect(x.get('/foo')).rejects.toEqual(
        new Error('xhr: request timed out')
      );
    });

    it('resolves if the readyState is 4 and the response is 200', () => {
      const x = Xhr(FakeHttpRequest);

      const promise = x.get('/some/resource');

      (xhrObj as any).readyState = 4;
      (xhrObj as any).responseText = 'working';
      (xhrObj as any).status = 200;
      (xhrObj as any).onload();

      return expect(promise).resolves.toEqual('working');
    });

    it('rejects if the readyState is 4 and the response is not 200 or 201', () => {
      const x = Xhr(FakeHttpRequest);

      const promise = x.get('/some/resource');

      (xhrObj as any).readyState = 4;
      (xhrObj as any).responseText = 'working';
      (xhrObj as any).status = 404;
      (xhrObj as any).onload();

      return expect(promise).rejects.toEqual(
        new Error('xhr: non 200 status: 404')
      );
    });

    it('resolves if the readyState is 4 and the response is 201', () => {
      const x = Xhr(FakeHttpRequest);

      const promise = x.get('/some/resource');

      (xhrObj as any).readyState = 4;
      (xhrObj as any).responseText = 'working';
      (xhrObj as any).status = 201;
      (xhrObj as any).onload();

      return expect(promise).resolves.toEqual('working');
    });

    it('defers until readyState is 4 and the response is 200', done => {
      const x = Xhr(FakeHttpRequest, 100, 1);
      let didRun = false;
      let didCheck = false;

      x.get('/some/resource')
        .then((value: string) => {
          didRun = true;
          expect(value).toBe('working');
          expect(didCheck).toBe(true);
          done();
        })
        .catch(done);

      // Trigger the test
      (xhrObj as any).readyState = 2;
      (xhrObj as any).onload();

      setTimeout(() => {
        didCheck = true;
        expect(didRun).toBe(false);

        // let the test pass
        (xhrObj as any).readyState = 4;
      }, 5);

      (xhrObj as any).status = 200;
      (xhrObj as any).responseText = 'working';
    });

    it('forwards headers', () => {
      const x = Xhr(FakeHttpRequest, 100, 1);
      const headers: { [key: string]: string } = {};
      xhrObj.setRequestHeader = (key: string, value: string) => {
        headers[key] = value;
      };
      x.get('/some/resource', undefined, { Authorization: 'bearer' });
      expect(headers.Authorization).toBe('bearer');
    });

    it('allows for default headers to be set', () => {
      const x = Xhr(FakeHttpRequest, 100, 1);
      const headers: { [key: string]: string } = {};
      xhrObj.setRequestHeader = (key: string, value: string) => {
        headers[key] = value;
      };
      x.defaultHeader('Authorization', 'bearer');
      x.get('/some/resource');
      expect(headers.Authorization).toBe('bearer');
    });

    it('allows for default headers to be unset', () => {
      const x = Xhr(FakeHttpRequest, 100, 1);
      let headers: { [key: string]: string } = {};
      xhrObj.setRequestHeader = (key: string, value: string) => {
        headers[key] = value;
      };
      x.defaultHeader('Authorization', 'bearer');
      x.get('/some/resource');
      expect(headers.Authorization).toBe('bearer');
      headers = {};
      x.defaultHeader('Authorization');
      x.get('/some/resource');
      expect(headers.Authorization).toBe(undefined);
    });

    it('allows for repeated calls to unset headers', () => {
      const x = Xhr(FakeHttpRequest, 100, 1);
      let headers: { [key: string]: string } = {};
      xhrObj.setRequestHeader = (key: string, value: string) => {
        headers[key] = value;
      };
      x.defaultHeader('Authorization', 'bearer');
      x.get('/some/resource');
      expect(headers.Authorization).toBe('bearer');
      headers = {};
      x.defaultHeader('Authorization');
      x.defaultHeader('Authorization');
      x.get('/some/resource');
      expect(headers.Authorization).toBe(undefined);
    });

    it('ignores falsey calls to defaultHeader', () => {
      const x = Xhr(FakeHttpRequest, 100, 1);
      const headers: { [key: string]: string } = {};
      xhrObj.setRequestHeader = (key: string, value: string) => {
        headers[key] = value;
      };
      x.defaultHeader('', 'bearer');
      x.get('/some/resource');
      expect(headers['']).toBe(undefined);
    });

    it('forces defaultHeader values to be strings', () => {
      const x = Xhr(FakeHttpRequest, 100, 1);
      const headers: { [key: string]: string } = {};
      xhrObj.setRequestHeader = (key: string, value: string) => {
        headers[key] = value;
      };
      x.defaultHeader('foo', 7 as any);
      x.get('/some/resource');
      expect(headers['foo']).toBe('7');
    });

    it('rejects put calls that have no data', () => {
      const x = Xhr(FakeHttpRequest);
      return expect(x.put('/somthing')).rejects.toEqual(
        new Error('xhr: newRequest: PUT: requires data')
      );
    });

    it('rejects post calls that have no data', () => {
      const x = Xhr(FakeHttpRequest);
      return expect(x.post('/somthing')).rejects.toEqual(
        new Error('xhr: newRequest: POST: requires data')
      );
    });

    it('stringifies post data', () => {
      const x = Xhr(FakeHttpRequest);
      const obj = { foo: 'bar' };
      let rec: any = null;
      xhrObj.send = (value: any) => {
        rec = value;
      };
      x.post('/somthing', obj);
      expect(rec).toEqual(JSON.stringify(obj));
    });

    it('rejects if something blows up synchronously', () => {
      const x = Xhr(FakeHttpRequest);
      const error = new Error('chaos!');
      xhrObj.send = (value: any) => {
        throw error;
      };
      return expect(x.get('/somthing')).rejects.toBe(error);
    });
  });

  describe('parseQuery function', () => {
    it('returns false if given no parameters', () => {
      expect(parseQuery()).toBe(false);
    });

    it('returns false if given a non object', () => {
      expect(parseQuery(7 as any)).toBe(false);
    });

    it('variablizes a single key/value pair', () => {
      const query = parseQuery({ foo: 'bar' });
      expect(query).toBe('?foo=bar');
    });

    it('variablizes a multiple key/value pairs', () => {
      const query = parseQuery({ foo: 'bar', baz: 'bof' });
      expect(query).toBe('?foo=bar&baz=bof');
    });
  });
});


src/xhr.ts
import { Dictionary, HttpHeaders, HttpMethod, IXhr } from './interfaces';

const json = 'application/json';
const DEFAULT_TIMEOUT = 190000;
const POLL_INTERVAL = 25;

export function Xhr(
  newXMLHTTPRequest?: () => XMLHttpRequest,
  defaultTimeout = DEFAULT_TIMEOUT,
  pollInterval = POLL_INTERVAL
): IXhr {
  return XhrConstructor(newXMLHTTPRequest as any, defaultTimeout, pollInterval);
}

function XhrConstructor(
  this: any,
  newXMLHTTPRequest: () => XMLHttpRequest,
  defaultTimeout: number,
  pollInterval: number
) {
  let XHRReq = newXMLHTTPRequest;
  // ensure object constructor
  if (!(this instanceof XhrConstructor)) {
    return new (<any>XhrConstructor)(
      newXMLHTTPRequest,
      defaultTimeout,
      pollInterval
    );
  }

  const that = this;
  /** @dict */
  const defaultHttpHeaders = Object.create(null);

  init();

  /**
   * @param {string} key
   * @param {string=} value
   * @returns {Object}
   */
  function defaultHeader(key: string, value?: string): HttpHeaders {
    if (!key) {
      return defaultHttpHeaders;
    }

    if (!value) {
      if (defaultHttpHeaders[key]) {
        delete defaultHttpHeaders[key];
      }
    } else {
      if (typeof value === 'string') {
        defaultHttpHeaders[key] = value;
      } else {
        defaultHttpHeaders[key] = value + '';
      }
    }

    return defaultHttpHeaders;
  }

  /**
   * @param {XMLHttpRequest} xhr
   */
  function setDefaultHttpHeaders(xhr: XMLHttpRequest) {
    Object.keys(defaultHttpHeaders).forEach(attr =>
      xhr.setRequestHeader(attr, defaultHttpHeaders[attr])
    );
  }

  /**
   * @param {XMLHttpRequest} xhr
   * @param {Object} localHttpHeaders
   */
  function setHttpHeaders(xhr: XMLHttpRequest, localHttpHeaders: HttpHeaders) {
    setDefaultHttpHeaders(xhr);

    Object.keys(localHttpHeaders).forEach(attr => {
      xhr.setRequestHeader(attr, localHttpHeaders[attr] + '');
    });
  }

  /**
   * @param {string} method
   * @param {string} url
   * @param {Object=} queryObj
   * @param {Object=} data
   * @param {string=} mimeType
   * @param {Object=} headers
   * @returns {Promise<string>}
   */
  function newRequest(
    method: HttpMethod,
    url: string,
    queryObj: Dictionary<any>,
    data: Dictionary<any>,
    mimeType: string,
    headers: HttpHeaders
  ): Promise<string> {
    return new Promise((resolve, reject) => {
      const finalMimeType = mimeType === undefined ? json : mimeType;
      const parsedQuery = parseQuery(queryObj);
      const finalUrl = parsedQuery ? url + parsedQuery : url;
      const xhr = (XHRReq as any)();
      let timer: any;

      if (method === 'POST' || method === 'PUT') {
        if (!data) {
          reject(new Error('xhr: newRequest: ' + method + ': requires data'));
          return;
        }
      }

      function clearTimer() {
        if (timer) {
          clearTimeout(timer);
          timer = 0;
        }
      }

      /**
       * @param {Error} reason
       */
      function onerror(reason: Error) {
        reject(reason);
        clearTimer();
      }

      function onload() {
        // if the document isn't ready try again later
        if (+xhr.readyState !== 4) {
          setTimeout(onload, pollInterval);
          return;
        }

        // if the header has a 200, or 201 status handle it
        if (xhr.status === 200 || xhr.status === 201) {
          resolve(xhr.responseText);
          clearTimer();
          return;
        }

        // otherwise error/unexpected
        onerror(new Error(`xhr: non 200 status: ${xhr.status}`));
      }

      try {
        /** @type {function(...)} */
        xhr.onload = onload;
        /** @type {function(...)} */
        xhr.onerror = onerror;
        xhr.open(method, finalUrl, true);
        xhr.setRequestHeader('Content-Type', finalMimeType);
        if (headers && typeof headers === 'object') {
          setHttpHeaders(xhr, headers);
        } else {
          setDefaultHttpHeaders(xhr);
        }

        if (!data) {
          xhr.send();
        } else {
          xhr.send(JSON.stringify(data));
        }
      } catch (e) {
        onerror(e);
        return;
      }

      timer = setTimeout(() => {
        if (xhr.abort) {
          xhr.abort();
        }
        onerror(new Error('xhr: request timed out'));
      }, defaultTimeout);
    });
  }

  /**
   * initialize the object
   */
  function init() {
    XHRReq =
      typeof newXMLHTTPRequest === 'function'
        ? newXMLHTTPRequest
        : () => new XMLHttpRequest();
    that['defaultHeader'] = defaultHeader;
    that['newRequest'] = newRequest;
  }
}

/**
 * @param {string} url
 * @param {Object=} queryObj
 * @param {Object=} headers
 * @returns {Promise<string>}
 */
XhrConstructor.prototype['get'] = function get(
  url: string,
  queryObj?: Dictionary<any>,
  headers?: HttpHeaders
) {
  return this['newRequest'](
    'GET',
    url,
    queryObj,
    undefined,
    undefined,
    headers
  );
};

/**
 * @param {string} url
 * @param {Object=} data
 * @param {Object=} queryObj
 * @param {Object=} headers
 * @returns {Promise<string>}
 */
XhrConstructor.prototype['put'] = function put(
  url: string,
  data?: Dictionary<any>,
  queryObj?: Dictionary<any>,
  headers?: HttpHeaders
) {
  return this['newRequest']('PUT', url, queryObj, data, undefined, headers);
};

/**
 * @param {string} url
 * @param {Object=} data
 * @param {Object=} queryObj
 * @param {Object=} headers
 * @returns {Promise<string>}
 */
XhrConstructor.prototype['post'] = function post(
  url: string,
  data?: Dictionary<any>,
  queryObj?: Dictionary<any>,
  headers?: HttpHeaders
) {
  return this['newRequest']('POST', url, queryObj, data, undefined, headers);
};

/**
 * @param {Object=} query
 * @returns {string|boolean}
 */
export function parseQuery(query?: Dictionary<any>): string | boolean {
  if (!query) {
    return false;
  }

  if (typeof query !== 'object') {
    return false;
  }

  let rString = '?';

  Object.keys(query).forEach((i, j) => {
    if (j !== 0) {
      rString += '&';
    }
    rString += [i, '=', encodeURIComponent(query[i])].join('');
  });

  return rString;
}


etc/karma.conf.js
// Karma configuration
// Generated on Tue Feb 04 2014 21:57:25 GMT-0500 (EST)

module.exports = function (config) {
    config.set(
        {
            // export CHROME_BIN=/usr/bin/chromium-browser
            // base path, that will be used to resolve files and exclude
            basePath: '..',

            // frameworks to use
            frameworks: ['jasmine'],


            // list of files / patterns to load in the browser
            files: [
                'lib/q/q.js',
                'lib/browser-fake-promise/fake-promise.js',

                'src/http.js',
                //'src/inline-xhr.js',

                'spec/mock*.js',
                'spec/mock*.js',
                'spec/*-spec.js',
                'spec/*Spec.js'
            ],


            // list of files to exclude
            exclude: [
            ],


            // test results reporter to use
            // possible values: 'dots', 'progress', 'junit', 'growl', 'coverage'
            reporters: ['progress', 'coverage'],


            // web server port
            port: 9876,


            // enable / disable colors in the output (reporters and logs)
            colors: true,


            // level of logging
            // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
            logLevel: config.LOG_INFO,


            // enable / disable watching file and executing tests whenever any file changes
            autoWatch: false,


            // Start these browsers, currently available:
            // - Chrome
            // - ChromeCanary
            // - Firefox
            // - Opera (has to be installed with `npm install karma-opera-launcher`)
            // - Safari (only Mac; has to be installed with `npm install karma-safari-launcher`)
            // - PhantomJS
            // - IE (only Windows; has to be installed with `npm install karma-ie-launcher`)
            browsers: ['Firefox', 'PhantomJS'], //, 'Chrome'],


            // If browser does not capture in given timeout [ms], kill it
            captureTimeout: 60000,


            // Continuous Integration mode
            // if true, it capture browsers, run tests and exit
            singleRun: false,

            // coverage support
            preprocessors: {
                'src/*.js': ['coverage']
            }
        });
};


src/inline-xhr.js
/**
 * quick xhr function
 *
 * @param url {string}
 * @param method {string}
 * @param data {*}
 * @param xsrf {string=} optional xsrf token
 * @param callback {function(...)}
 */
function fetch(url, method, data, xsrf, callback) {
    'use strict';
    if (typeof callback !== 'function') {
        callback = function () {
        };
    }
    /** @type {!Object} */
    var xhr = new XMLHttpRequest(),
    noResponseTimer = setTimeout(function () {
        xhr.abort();
        callback(new Error('timeout'));
    }, 60000);

    if (typeof callback !== 'function') {
        callback = function () {
        };
    }
    xhr.onreadystatechange = function (e) {
        if (xhr.readyState !== 4) {
            return;
        }
        if (xhr.status === 200) {
            clearTimeout(noResponseTimer);
            callback(null, xhr.responseText);
        } else {
            callback(new Error('connection error'));
        }
    };
    xhr.open(method, url);
    if (xsrf) {
        xhr.setRequestHeader('X-XSRF-TOKEN', xsrf);
    }
    xhr.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded; charset=UTF-8');
    if (!data) {
        xhr.send();
    } else {
        xhr.send(data);
    }
}


src/interfaces.ts
import { Dictionary } from '@ch1/utility';
export { Dictionary } from '@ch1/utility';

export type HttpHeaders = Dictionary<string>;
export type HttpMethod = 'DELETE' | 'GET' | 'POST' | 'PUT';

export interface IXhr {
  defaultHeader(key: string, value?: string): HttpHeaders;
  get(
    url: string,
    queryObj?: Dictionary<any>,
    headers?: HttpHeaders
  ): Promise<string>;
  newRequest(
    method: HttpMethod,
    url: string,
    queryObj?: Dictionary<any>,
    data?: Dictionary<any>,
    mimeType?: string,
    headers?: HttpHeaders
  ): Promise<string>;
  post(
    url: string,
    data?: Dictionary<any>,
    queryObj?: Dictionary<any>,
    headers?: HttpHeaders
  ): Promise<string>;
  put(
    url: string,
    data?: Dictionary<any>,
    queryObj?: Dictionary<any>,
    headers?: HttpHeaders
  ): Promise<string>;
}


src/http-client.js
/*global workular, initLog */

workular.module('js-http', []).factory('stringsHTTP', [function () {
    'use strict';

    /** @type {Object} */
    var S = Object.create(null);

    /** @type {string} */
    S['self'] = 'HOMER.http';
    /** @type {string} */
    S['query'] = '?';
    /** @type {string} */
    S['concat'] = '&';
    /** @type {string} */
    S['assign'] = '=';
    /** @type {string} */
    S['fnOnload'] = 'onload';
    /** @type {string} */
    S['fnOnerror'] = 'onerror';
    /** @type {string} */
    S['fnOnprogress'] = 'onprogress';
    /** @type {string} */
    S['status'] = 'status code';
    /** @type {string} */
    S['timeout'] = 'request timed out';

    /** @type {string} */
    S['headerContent'] = 'Content-Type';
    /** @type {string} */
    S['get'] = 'GET';
    /** @type {string} */
    S['put'] = 'PUT';
    /** @type {string} */
    S['post'] = 'POST';
    /** @type {string} */
    S['Object'] = 'object';
    /** @type {string} */
    S['dirSlash'] = '/';

    /** @const */
    Object.freeze(S);

    return S;
}]).factory('newXMLHttpRequest', ['global', function () {
    'use strict';
    /**
     * allows for XHR mocks
     *
     * @return {!Object}
     */
    function newXMLHttpRequest() {
        return new XMLHttpRequest();
    }

    return newXMLHttpRequest;
}]).factory('http', ['q', 'stringsHTTP', 'log', 'newXMLHttpRequest', function (Q, S, log, newXMLHttpRequest) {
    'use strict';


    return http;
}]).factory('inlineXHR', [function () {
    'use strict';

    return fetch;
}]);

src/http.js
/**
 * Created by michael on 23/05/14.
 */
/*global SimpleFakePromise */

function HTTP(newXMLHTTPRequest) {
    'use strict';

    // ensure object constructor
    if (!(this instanceof HTTP)) {
        return new HTTP(newXMLHTTPRequest);
    }

    var that = this,
        log,
        Q = new SimpleFakePromise(),
        /** @dict */
        consts = Object.freeze({
            json: 'application/json',
            noop: function () {
            }
        }),
        defaultHeaders = Object.create(null),
        /** @type Array.<function(...)> */
        noAccessHooks = [],
        /** @type {boolean} default to true */
        isOnline = true,
        /** @type {number} */
        defaultTimeout = 190000;

    /*global console*/
    if (typeof console !== 'undefined') {
        log = console;
    } else {
        log = {
            log: consts.noop,
            info: consts.noop,
            warn: consts.noop,
            error: consts.noop,
            assert: consts.noop
        };
    }

    /**
     * @param fn {function (...)}
     * @returns {boolean}
     */
    function isFunction(fn) {
        return typeof fn === 'function';
    }

    /**
     * upgrades the logger
     * @param newLog
     */
    function setLogger(newLog) {
        if (isFunction(newLog.log) && isFunction(newLog.info) && isFunction(newLog.warn) && isFunction(newLog.error) && isFunction(newLog.assert)) {
            Object.keys(newLog).forEach(function (newLogAttr) {
                log[newLogAttr] = newLog[newLogAttr];
            });
        }
    }

    /**
     * Sets JS RPC to use promises instead of callbacks, using the given promise
     * library
     * @param lib {Object} a promise library like Q
     * @returns {boolean}
     *
     * Promise library must support:
     * lib.defer()
     * lib.defer().resolve()
     * lib.defer().reject()
     * lib.defer().promise
     * lib.defer().promise.then()
     */
    function setPromiseLib(lib) {
        if (!isFunction(lib.defer)) {
            throw new TypeError('RPC: setPromiseLib: expecting a library with a root level defer function');
        }
        var test = lib.defer();

        if (!isFunction(test.resolve)) {
            throw new TypeError('RPC: setPromiseLib: expecting defers to have a resolve method');
        }

        if (!isFunction(test.reject)) {
            throw new TypeError('RPC: setPromiseLib: expecting defers to have a reject method');
        }

        if (!test.promise) {
            throw new TypeError('RPC: setPromiseLib: expecting defer objects to have a promise');
        }

        if (!isFunction(test.promise.then)) {
            throw new TypeError('RPC: setPromiseLib: expecting promises to have a then method');
        }

        HTTP.prototype['Q'] = lib;
        Q = lib;
        return true;
    }


    /**
     * adds a default header to be used on http requests
     * @param key
     * @param value
     * @returns {boolean}
     */
    function defaultHeader(key, value) {
        if (!key) {
            return defaultHeaders;
        }
        if ((typeof key !== 'string') && (typeof key !== 'string')) {
            return false;
        }
        if (value === '') {
            if (defaultHeaders[key]) {
                delete defaultHeaders[key];
            }
        } else {
            defaultHeaders[key] = value;
        }
        return true;
    }

    /**
     *
     * @param fn (function (...))
     */
    function addNoAccess(fn) {
        if (!isFunction(fn)) {
            return false;
        }
        noAccessHooks.push(fn);
        return true;
    }

    /**
     * @param {number} code - HTTP response code
     * triggers the items in noAccess
     */
    function onNoAccess(code) {
        noAccessHooks.forEach(function (hook) {
            try {
                hook(code);
            } catch (err) {
                // fail silent
            }
        });
    }

    /**
     * assigns the default headers to an xhr object
     * @param xhr {Object}
     */
    function setDefaultHeaders(xhr) {
        Object.keys(defaultHeaders).forEach(function (attr) {
            xhr.setRequestHeader(attr, defaultHeaders[attr]);
        });
    }

    /**
     * assigns the default headers, then assigns given headers
     * @param xhr {Object}
     * @param localHeaders {Object}
     */
    function setHeaders(xhr, localHeaders) {
        setDefaultHeaders(xhr);

        Object.keys(localHeaders).forEach(function (attr) {
            if (typeof localHeaders[attr] !== 'string') {
                localHeaders[attr] = localHeaders[attr].toString();
            }
            xhr.setRequestHeader(attr, localHeaders[attr]);
        });
    }

    /**
     *
     * @param method {string}
     * @param url {string}
     * @param queryObj {Object=}
     * @param data {*}
     * @param mimeType {string=}
     * @returns {*}
     */
    function newRequest(method, url, queryObj, data, mimeType, headers) {
        if (isOnline === false) {
            var d = Q.defer();
            d.reject(new Error('http: offline'));
            return d.promise;
        }

        mimeType = (mimeType === undefined) ? consts.json : mimeType;
        queryObj = that.parseQuery(queryObj);
        if (queryObj !== false) {
            url += queryObj;
        }

        /** @type {!Object} */
        var xhr = newXMLHTTPRequest(),
            /** @type {!Object} */
            deferred = Q.defer(),
            timer = false;

        if ((method === 'POST') || (method === 'PUT')) {
            if (!data) {
                deferred.reject(new Error('http: newRequest: ' + method + ': has falsy data'));
                return deferred.promise;
            }
        }

        function clearTimer() {
            if (timer !== false) {
                clearTimeout(timer);
                timer = false;
            }
        }

        function onerror(reason) {
            reason = (!reason) ? 'fatal error' : reason;

            deferred.reject(new Error(['http',
                reason,
                JSON.stringify(url)].join(':')));
            clearTimer();
        }

        function onload() {
            // if the document isn't ready try again later
            if (+xhr.readyState !== 4) {
                return setTimeout(onload, 25);
            }
            // if the header has a 200, or 201 status handle it
            if ((xhr.status === 200) || (xhr.status === 201)) {
                deferred.resolve(xhr.responseText);
                clearTimer();
                return;
            }
            // if the header has a 400/401 status trip the logout flag
            if (( xhr.status === 401) || (xhr.status === 400)) {
                onNoAccess(+xhr.status);
            }
            // otherwise error/unexpected
            onerror('status code');
        }

        function onprogress(readyState) {
            if (readyState) {
                if (readyState.lengthComputable) {
                    if (typeof deferred.notify === 'function') {
                        deferred.notify(readyState);
                    }
                }
            }
        }

        try {
            /** @type {function(...)} */
            xhr.onload = onload;
            /** @type {function(...)} */
            xhr.onerror = function () {
                onerror('http error');
            };
            /** @type {function(...)} */
            xhr.onprogress = onprogress;

            xhr.open(method, that.host + url, true);
            xhr.setRequestHeader('Content-Type', mimeType);
            if ((headers) && (typeof headers === 'object')) {
                setHeaders(xhr, headers);
            } else {
                setDefaultHeaders(xhr);
            }

            if ((data === null) ||
                (data === false) ||
                (data === undefined)) {
                xhr.send();
            } else {
                xhr.send(JSON.stringify(data));
            }
        } catch (e) {
            onerror('unhandled exception ' + e.message);
            return deferred.promise;
        }

        timer = setTimeout(function () {
            var errorstr = 'request timed out';

            try {
                xhr.abort();
                onerror(errorstr);
            } catch (err) {
                errorstr = err.message;
                onerror(errorstr);
            }
        }, defaultTimeout);

        return deferred.promise;
    }

    /**
     * gets/sets the default timeout
     * @param val {number}
     * @returns {number}
     */
    function timeout(val) {
        if (!val) { return defaultTimeout; }
        val = val.toString();
        val = +val;
        if (!val) { return defaultTimeout; }
        if ((val < 2500) || (val > 600000)) { return defaultTimeout; }
        defaultTimeout = +val;
        return defaultTimeout;
    }

    /**
     * gets/sets the online status of this object.
     * @param val {boolean} truthy sets isOnline to true, false sets isOnline to false
     * @returns {boolean}
     */
    function online(val) {
        if (val) {
            isOnline = true;
        }
        if (val === false) {
            isOnline = false;
        }
        return isOnline;
    }

    /**
     * expose the api
     */
    function expose() {
        that['setLogger'] = setLogger;
        that['setPromiseLib'] = setPromiseLib;
        that['defaultHeader'] = defaultHeader;
        that['onNoAccess'] = addNoAccess;
        that['newRequest'] = newRequest;
        that['timeout'] = timeout;
        that['online'] = online;
        that['host'] = '';
        that['log'] = log;
    }

    /**
     * initialize the object
     */
    function init() {
        newXMLHTTPRequest = typeof newXMLHTTPRequest === 'function' ? newXMLHTTPRequest : function () {
            return new XMLHttpRequest(Array.prototype.slice.call(arguments, 0));
        };
        expose();
    }

    init();
}

/**
 * @param {string} url
 * @param {Object=} queryObj
 * @param {Object=} headers
 *
 * @return {!Object}
 */
HTTP.prototype['get'] = function get(url, queryObj, headers) {
    'use strict';
    return this.newRequest('GET', url, queryObj, undefined, undefined, headers);
};

/**
 * @param {string} url
 * @param {*} data
 * @param {Object=} queryObj
 * @param {Object=} headers
 *
 * @return {!Object}
 */
HTTP.prototype['put'] = function put(url, data, queryObj, headers) {
    'use strict';
    return this.newRequest('PUT', url, queryObj, data, undefined, headers);
};

/**
 * @param {string} url
 * @param {*} data
 * @param {!Object=} queryObj
 *
 * @return {!Object}
 */
HTTP.prototype['post'] = function post(url, data, queryObj, headers) {
    'use strict';
    return this.newRequest('POST', url, queryObj, data, undefined, headers);
};

/**
 * @param {string=} query
 *
 * @return {string|boolean}
 */
HTTP.prototype['parseQuery'] = function parseQuery(query) {
    'use strict';

    if (!query) {
        return false;
    }

    if (typeof query !== 'object') {
        return false;
    }

    var j = 0, rString = '?';

    try {

        Object.keys(query).forEach(function (i) {
            if (j !== 0) {
                rString += '&';
            }
            j += 1;
            rString += [i, '=', encodeURIComponent(query[i])].join('');
        });
    } catch (err) {
        this.log.warn('http: parseQuery: ' + err.message);
        return '';
    }
    return rString;
};

HTTP.prototype['Q'] = new SimpleFakePromise();
